File: scripts/upload_drive.js
#!/usr/bin/env node
/**
 * Upload files to Google Drive "Tiktok ready" folder.
 * Extracts OAuth access token from Chrome's active session via CDP,
 * then uses Google Drive REST API to upload files directly.
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const https = require('https');

const CDP_PORT = 9222;
const FOLDER_ID = '1hGKOtWbET3cmPMZ6zK85-heFQRq9s5k8'; // "Tiktok ready" folder
const OVERLAID_DIR = path.resolve(__dirname, '..', 'output', 'gemgen_batch', 'overlaid');

async function getAccessToken(browser) {
  const contexts = browser.contexts();
  let page = null;
  
  // Find or open a Google page
  for (const ctx of contexts) {
    for (const p of ctx.pages()) {
      if (p.url().includes('google.com')) {
        page = p;
        break;
      }
    }
    if (page) break;
  }
  
  if (!page) {
    page = await contexts[0].newPage();
    await page.goto('https://drive.google.com', { waitUntil: 'networkidle', timeout: 30000 });
    await page.waitForTimeout(3000);
  }

  // Navigate to the OAuth token endpoint
  // Google's internal token endpoint returns a valid access token for the logged-in user
  const tokenPage = await contexts[0].newPage();
  await tokenPage.goto('https://accounts.google.com/o/oauth2/auth?' + 
    'client_id=65746573420-jcjmpgep677t35h8n03nl0hhk7t60hk3.apps.googleusercontent.com' +
    '&redirect_uri=https://www.googleapis.com/auth/callback' +
    '&scope=https://www.googleapis.com/auth/drive.file' +
    '&response_type=token' +
    '&prompt=none', { timeout: 15000 }).catch(() => {});
  
  await tokenPage.waitForTimeout(3000);
  
  // Try to extract token from URL fragment
  const url = tokenPage.url();
  const hashMatch = url.match(/access_token=([^&]+)/);
  if (hashMatch) {
    await tokenPage.close();
    return hashMatch[1];
  }

  // Fallback: Use the Drive API through the browser's cookies
  // Extract a token by making XHR request from Drive page
  const token = await page.evaluate(async () => {
    try {
      const resp = await fetch('https://www.googleapis.com/drive/v3/about?fields=user', {
        credentials: 'include'
      });
      // If this works, we have cookie-based auth
      if (resp.ok) return '__cookie_auth__';
    } catch (e) {}
    return null;
  });

  await tokenPage.close().catch(() => {});
  return token;
}

function uploadFile(filePath, accessToken) {
  return new Promise((resolve, reject) => {
    const fileName = path.basename(filePath);
    const fileSize = fs.statSync(filePath).size;
    
    // Step 1: Create file metadata
    const metadata = JSON.stringify({
      name: fileName,
      parents: [FOLDER_ID]
    });
    
    const boundary = '-------314159265358979323846';
    const delimiter = `\r\n--${boundary}\r\n`;
    const closeDelimiter = `\r\n--${boundary}--`;
    
    const fileData = fs.readFileSync(filePath);
    
    const multipartBody = Buffer.concat([
      Buffer.from(delimiter),
      Buffer.from('Content-Type: application/json; charset=UTF-8\r\n\r\n'),
      Buffer.from(metadata),
      Buffer.from(delimiter),
      Buffer.from('Content-Type: image/png\r\n\r\n'),
      fileData,
      Buffer.from(closeDelimiter)
    ]);
    
    const options = {
      hostname: 'www.googleapis.com',
      path: '/upload/drive/v3/files?uploadType=multipart',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${accessToken}`,
        'Content-Type': `multipart/related; boundary=${boundary}`,
        'Content-Length': multipartBody.length
      }
    };
    
    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 200 || res.statusCode === 201) {
          const result = JSON.parse(data);
          resolve(result);
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data.substring(0, 200)}`));
        }
      });
    });
    
    req.on('error', reject);
    req.write(multipartBody);
    req.end();
  });
}

async function uploadViaBrowser(browser, filePath) {
  // Alternative: use browser's authenticated session to upload via fetch
  const contexts = browser.contexts();
  let page = null;
  
  for (const ctx of contexts) {
    for (const p of ctx.pages()) {
      if (p.url().includes('drive.google.com')) {
        page = p;
        break;
      }
    }
    if (page) break;
  }
  
  if (!page) {
    page = await contexts[0].newPage();
    await page.goto('https://drive.google.com', { waitUntil: 'networkidle', timeout: 30000 });
    await page.waitForTimeout(3000);
  }

  const fileName = path.basename(filePath);
  const fileData = fs.readFileSync(filePath);
  const base64Data = fileData.toString('base64');
  
  const result = await page.evaluate(async ({ fileName, base64Data, folderId }) => {
    try {
      // Convert base64 back to binary
      const binaryStr = atob(base64Data);
      const bytes = new Uint8Array(binaryStr.length);
      for (let i = 0; i < binaryStr.length; i++) {
        bytes[i] = binaryStr.charCodeAt(i);
      }
      const blob = new Blob([bytes], { type: 'image/png' });
      
      const metadata = {
        name: fileName,
        parents: [folderId]
      };
      
      const formData = new FormData();
      formData.append('metadata', new Blob([JSON.stringify(metadata)], { type: 'application/json' }));
      formData.append('file', blob);
      
      const resp = await fetch('https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart', {
        method: 'POST',
        body: formData
      });
      
      if (resp.ok) {
        const data = await resp.json();
        return { success: true, id: data.id, name: data.name };
      } else {
        const text = await resp.text();
        return { success: false, error: `HTTP ${resp.status}: ${text.substring(0, 200)}` };
      }
    } catch (err) {
      return { success: false, error: err.message };
    }
  }, { fileName, base64Data, folderId: FOLDER_ID });
  
  return result;
}

async function main() {
  const files = fs.readdirSync(OVERLAID_DIR)
    .filter(f => /\.(png|jpg|jpeg)$/i.test(f))
    .sort()
    .map(f => path.join(OVERLAID_DIR, f));

  if (files.length === 0) {
    console.error('❌ No images found.');
    process.exit(1);
  }

  console.log(`Found ${files.length} images to upload.\n`);

  console.log(`Connecting to Chrome on CDP port ${CDP_PORT}...`);
  const browser = await chromium.connectOverCDP(`http://localhost:${CDP_PORT}`);
  
  // Try OAuth token approach first
  console.log('Extracting access token...');
  const token = await getAccessToken(browser);
  
  let uploaded = 0;
  let failed = 0;

  if (token && token !== '__cookie_auth__') {
    console.log('Got OAuth token. Using direct API upload.\n');
    
    for (const file of files) {
      const basename = path.basename(file);
      const sizeMB = (fs.statSync(file).size / (1024 * 1024)).toFixed(1);
      process.stdout.write(`  [${uploaded + failed + 1}/${files.length}] ${basename} (${sizeMB}MB)... `);
      
      try {
        const result = await uploadFile(file, token);
        console.log(`✅ (id: ${result.id})`);
        uploaded++;
      } catch (err) {
        console.log(`❌ ${err.message.substring(0, 80)}`);
        failed++;
      }
    }
  } else {
    console.log('No direct token. Using browser-authenticated upload.\n');
    
    for (const file of files) {
      const basename = path.basename(file);
      const sizeMB = (fs.statSync(file).size / (1024 * 1024)).toFixed(1);
      process.stdout.write(`  [${uploaded + failed + 1}/${files.length}] ${basename} (${sizeMB}MB)... `);
      
      try {
        const result = await uploadViaBrowser(browser, file);
        if (result.success) {
          console.log(`✅ (id: ${result.id})`);
          uploaded++;
        } else {
          console.log(`❌ ${result.error}`);
          failed++;
        }
      } catch (err) {
        console.log(`❌ ${err.message.substring(0, 80)}`);
        failed++;
      }
    }
  }

  console.log(`\n========== RESULTS ==========`);
  console.log(`Uploaded: ${uploaded}/${files.length}`);
  console.log(`Failed: ${failed}`);
}

main().catch(err => {
  console.error('Fatal error:', err.message);
  process.exit(1);
});

File: scripts/upload_drive.py
"""
Upload overlaid images + TikTok metadata to Google Drive "Tiktok ready" folder.
Uses Google Drive API (OAuth 2.0).

Flags:
  --clean         Delete all existing files in Drive folder before uploading
  --dir PATH      Override upload directory (default: output/gemgen_batch/overlaid)
  --include-raw   Also upload uncaptioned images from gemgen_batch/ as wallpapers
  --pinterest     Upload raw images to a PUBLIC Drive folder for Pinterest bulk CSV.
                  Generates pinterest_urls.json with {filename: public_url} mapping.
"""

import os
import sys
import json
import glob
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload

SCOPES = ['https://www.googleapis.com/auth/drive.file']
BASE_DIR = os.path.join(os.path.dirname(__file__), '..')
CREDS_FILE = os.path.join(BASE_DIR, 'credentials.json')
TOKEN_FILE = os.path.join(BASE_DIR, 'token.json')
FOLDER_ID = '1hGKOtWbET3cmPMZ6zK85-heFQRq9s5k8'  # "Tiktok ready" folder
PINTEREST_FOLDER_NAME = 'Pinterest Bulk'  # Auto-created public folder
OVERLAID_DIR = os.path.join(BASE_DIR, 'output', 'gemgen_batch', 'overlaid')
RAW_DIR = os.path.join(BASE_DIR, 'output', 'gemgen_batch')
PINTEREST_URLS_FILE = os.path.join(BASE_DIR, 'pinterest_urls.json')

MIME_MAP = {
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.txt': 'text/plain',
}


def get_drive_service():
    """Authenticate and return Drive API service."""
    creds = None
    if os.path.exists(TOKEN_FILE):
        creds = Credentials.from_authorized_user_file(TOKEN_FILE, SCOPES)

    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            print("Refreshing expired token...")
            creds.refresh(Request())
        else:
            print("Starting local server for authentication...")
            flow = InstalledAppFlow.from_client_secrets_file(CREDS_FILE, SCOPES)
            creds = flow.run_local_server(port=0)
        with open(TOKEN_FILE, 'w') as f:
            f.write(creds.to_json())
        print("Token saved.\n")

    return build('drive', 'v3', credentials=creds)


def list_drive_files(service, folder_id):
    """List all files in the Drive folder. Returns list of {id, name, size}."""
    results = []
    page_token = None
    while True:
        resp = service.files().list(
            q=f"'{folder_id}' in parents and trashed=false",
            fields='nextPageToken, files(id, name, size)',
            pageToken=page_token
        ).execute()
        results.extend(resp.get('files', []))
        page_token = resp.get('nextPageToken')
        if not page_token:
            break
    return results


def clean_drive_folder(service, folder_id):
    """Delete all files in the Drive folder (#2: auto-clean before upload)."""
    existing = list_drive_files(service, folder_id)
    if not existing:
        print("  Drive folder already empty.\n")
        return 0

    print(f"  Deleting {len(existing)} existing file(s)...")
    deleted = 0
    for f in existing:
        try:
            service.files().delete(fileId=f['id']).execute()
            deleted += 1
        except Exception as e:
            print(f"    ⚠️  Could not delete {f['name']}: {e}")
    print(f"  Cleared {deleted} file(s).\n")
    return deleted


def upload_file(service, filepath, folder_id, existing_files=None):
    """Upload a file. Skips if identical file exists (#6: dedup)."""
    filename = os.path.basename(filepath)
    local_size = str(os.path.getsize(filepath))

    # Dedup check
    if existing_files:
        for ef in existing_files:
            if ef['name'] == filename and ef.get('size') == local_size:
                return None  # Skip — identical file exists

    ext = os.path.splitext(filepath)[1].lower()
    mimetype = MIME_MAP.get(ext, 'application/octet-stream')

    file_metadata = {'name': filename, 'parents': [folder_id]}
    media = MediaFileUpload(filepath, mimetype=mimetype, resumable=True)

    result = service.files().create(
        body=file_metadata,
        media_body=media,
        fields='id, name'
    ).execute()
    return result


def collect_raw_images(raw_dir):
    """Collect raw (uncaptioned) PNGs from gemgen_batch/, excluding subdirectories."""
    raw_files = sorted(glob.glob(os.path.join(raw_dir, '*.png')))
    if not raw_files:
        raw_files = sorted(glob.glob(os.path.join(raw_dir, '*.jpg')))
    return raw_files


# ── Pinterest Bulk Upload Functions ──────────────────────────────────────────

def get_or_create_pinterest_folder(service):
    """Find or create the 'Pinterest Bulk' folder in Drive root. Returns folder ID."""
    # Search for existing folder
    resp = service.files().list(
        q=f"name='{PINTEREST_FOLDER_NAME}' and mimeType='application/vnd.google-apps.folder' and trashed=false",
        fields='files(id, name)',
        spaces='drive'
    ).execute()
    existing = resp.get('files', [])
    if existing:
        folder_id = existing[0]['id']
        print(f"  Found existing '{PINTEREST_FOLDER_NAME}' folder: {folder_id}")
        return folder_id

    # Create new folder
    metadata = {
        'name': PINTEREST_FOLDER_NAME,
        'mimeType': 'application/vnd.google-apps.folder'
    }
    folder = service.files().create(body=metadata, fields='id').execute()
    folder_id = folder['id']
    print(f"  Created '{PINTEREST_FOLDER_NAME}' folder: {folder_id}")

    # Make folder itself publicly viewable
    service.permissions().create(
        fileId=folder_id,
        body={'type': 'anyone', 'role': 'reader'},
        fields='id'
    ).execute()
    print(f"  Folder set to public (anyone with link)")
    return folder_id


def make_file_public(service, file_id):
    """Set a Drive file to 'anyone with link can view'."""
    service.permissions().create(
        fileId=file_id,
        body={'type': 'anyone', 'role': 'reader'},
        fields='id'
    ).execute()


def get_direct_url(file_id):
    """Get a direct-access URL for a Drive file (for Pinterest CSV)."""
    # Primary: lh3 proxy (fast, no redirect, works for images)
    # Fallback in CSV comments: drive.google.com/uc?export=view&id=...
    return f"https://lh3.googleusercontent.com/d/{file_id}"


def run_pinterest_mode(service, source_dir=None):
    """Upload images to public Drive folder and generate pinterest_urls.json."""
    print("=" * 50)
    print("  Pinterest Bulk Upload — Public Image Hosting")
    print("=" * 50 + "\n")

    # Determine source directory
    if not source_dir:
        source_dir = RAW_DIR

    # Collect images
    files = sorted(glob.glob(os.path.join(source_dir, '*.png')))
    if not files:
        files = sorted(glob.glob(os.path.join(source_dir, '*.jpg')))
    if not files:
        print(f"No images found in {source_dir}")
        sys.exit(1)

    print(f"Found {len(files)} images to upload:\n")
    total_size = 0
    for f in files:
        size_mb = os.path.getsize(f) / (1024 * 1024)
        total_size += size_mb
        print(f"  {os.path.basename(f)} — {size_mb:.1f}MB")
    print(f"\n  Total: {total_size:.1f}MB\n")

    # Get or create the Pinterest folder
    print("Setting up Pinterest folder...")
    pinterest_folder_id = get_or_create_pinterest_folder(service)

    # Clean existing files in Pinterest folder
    existing = list_drive_files(service, pinterest_folder_id)
    if existing:
        print(f"\n  Clearing {len(existing)} old file(s) from Pinterest folder...")
        clean_drive_folder(service, pinterest_folder_id)

    # Upload each image, make public, collect URLs
    url_map = {}  # {filename: {id, url, drive_url}}
    uploaded = 0
    failed = 0

    print(f"\nUploading {len(files)} images...\n")
    for i, filepath in enumerate(files):
        basename = os.path.basename(filepath)
        size_mb = os.path.getsize(filepath) / (1024 * 1024)
        print(f"  [{i+1}/{len(files)}] {basename} ({size_mb:.1f}MB)...", end=" ", flush=True)

        try:
            result = upload_file(service, filepath, pinterest_folder_id)
            if result:
                file_id = result['id']
                # Make publicly accessible
                make_file_public(service, file_id)
                direct_url = get_direct_url(file_id)
                url_map[basename] = {
                    'id': file_id,
                    'url': direct_url,
                    'drive_url': f"https://drive.google.com/uc?export=view&id={file_id}"
                }
                print(f"OK")
                uploaded += 1
            else:
                print("SKIP (duplicate)")
        except Exception as e:
            print(f"FAIL: {str(e)[:80]}")
            failed += 1

    # Write URL mapping
    with open(PINTEREST_URLS_FILE, 'w', encoding='utf-8') as f:
        json.dump(url_map, f, indent=2)

    print(f"\n{'=' * 50}")
    print(f"  Uploaded: {uploaded} | Failed: {failed}")
    print(f"  URL map saved to: {PINTEREST_URLS_FILE}")
    print(f"{'=' * 50}\n")

    # Print sample URLs for verification
    if url_map:
        sample_key = list(url_map.keys())[0]
        sample = url_map[sample_key]
        print(f"Sample URL (paste in browser to test):")
        print(f"  {sample['url']}\n")

    return url_map




def main():
    args = sys.argv[1:]
    do_clean = '--clean' in args
    if do_clean:
        args.remove('--clean')

    include_raw = '--include-raw' in args
    if include_raw:
        args.remove('--include-raw')

    pinterest_mode = '--pinterest' in args
    if pinterest_mode:
        args.remove('--pinterest')

    upload_dir = OVERLAID_DIR
    if '--dir' in args:
        idx = args.index('--dir')
        upload_dir = args[idx + 1]

    # Authenticate first (shared across modes)
    service = get_drive_service()

    # Pinterest mode: upload to public folder + generate URL mapping
    if pinterest_mode:
        source_dir = upload_dir if '--dir' in sys.argv else None
        run_pinterest_mode(service, source_dir)
        return

    print("=" * 50)
    print("  Google Drive Upload — TrendTok Pipeline")
    print("=" * 50 + "\n")

    # Find overlaid upload files
    files = sorted(glob.glob(os.path.join(upload_dir, '*.png')))
    if not files:
        files = sorted(glob.glob(os.path.join(upload_dir, '*.jpg')))
    if not files:
        print(f"❌ No images found in {upload_dir}")
        sys.exit(1)

    # Collect raw wallpaper images if requested
    raw_files = []
    if include_raw:
        raw_files = collect_raw_images(RAW_DIR)
        if raw_files:
            print(f"🖼️  --include-raw: Found {len(raw_files)} uncaptioned wallpaper images\n")
        else:
            print("ℹ️  --include-raw: No raw images found in gemgen_batch/\n")

    # Check for metadata
    metadata_txt = os.path.join(upload_dir, 'tiktok_metadata.txt')
    if not os.path.exists(metadata_txt):
        metadata_txt = None
        print("ℹ️  No tiktok_metadata.txt found — uploading images only\n")
    else:
        print("📝 Found tiktok_metadata.txt — will upload alongside images\n")

    # List overlaid files
    total_size = 0
    print(f"Found {len(files)} overlaid (captioned) images to upload:\n")
    for f in files:
        size_mb = os.path.getsize(f) / (1024 * 1024)
        total_size += size_mb
        print(f"  📝 {os.path.basename(f)} — {size_mb:.1f}MB")

    # List raw wallpaper files
    if raw_files:
        print(f"\nFound {len(raw_files)} raw (wallpaper) images to upload:\n")
        for f in raw_files:
            size_mb = os.path.getsize(f) / (1024 * 1024)
            total_size += size_mb
            print(f"  🖼️  {os.path.basename(f)} — {size_mb:.1f}MB")

    if metadata_txt:
        print(f"  tiktok_metadata.txt — {os.path.getsize(metadata_txt) / 1024:.1f}KB")
    print(f"\n  Total: {total_size:.1f}MB + metadata\n")


    # Clean old files if requested (#2)
    existing_files = []
    if do_clean:
        print("🧹 --clean: Clearing Drive folder before upload...")
        clean_drive_folder(service, FOLDER_ID)
    else:
        # Load existing for dedup (#6)
        existing_files = list_drive_files(service, FOLDER_ID)
        if existing_files:
            print(f"📂 {len(existing_files)} file(s) already in Drive — will skip duplicates\n")

    # Upload
    uploaded = 0
    skipped = 0
    failed = 0

    # Build full file list: overlaid + raw wallpapers + metadata
    all_files = files + raw_files + ([metadata_txt] if metadata_txt else [])
    total_files = len(all_files)

    for i, filepath in enumerate(all_files):
        basename = os.path.basename(filepath)
        size_mb = os.path.getsize(filepath) / (1024 * 1024)

        if filepath == metadata_txt:
            label = "META"
        elif filepath in raw_files:
            label = f"RAW {raw_files.index(filepath)+1}/{len(raw_files)}"
        else:
            label = f"{files.index(filepath)+1}/{len(files)}"

        print(f"[{label}] {basename} ({size_mb:.1f}MB)...", end=" ", flush=True)

        try:
            result = upload_file(service, filepath, FOLDER_ID,
                                 existing_files if not do_clean else None)
            if result is None:
                print("⏭️  (duplicate, skipped)")
                skipped += 1
            else:
                print(f"✅ (id: {result['id']})")
                uploaded += 1
        except Exception as e:
            print(f"❌ {str(e)[:100]}")
            failed += 1

    print(f"\n{'=' * 50}")
    print(f"Uploaded: {uploaded} | Skipped: {skipped} | Failed: {failed} | Total: {total_files}")
    if include_raw and raw_files:
        print(f"  (includes {len(raw_files)} uncaptioned wallpaper images)")
    print(f"{'=' * 50}")


if __name__ == '__main__':
    main()

